Issue #191: Allow leeway to handle clock skew

m777z opened the pull request:
Not sure if I'm doing this right since it's my first real code contribution to an open-source project. I'm sending the PR to 3.3 per @lcobucci's comment on this issue on May 17. 

Note that all the whitespace changes were made by my editor (PHPStorm) when set to the "PSR1/PSR2" predefined style.

Ocramius commented on README.md:
```diff
@@ -88,6 +88,27 @@ var_dump($token->validate($data)); // true, because current time is between "nbf
 
 var_dump($token->validate($data)); // false, because token is expired since current time is greater than exp
+
+// We can also use leeway to deal with clock skew.
```
I think most readers of the `README.md` will unlikely know what `leeway` or `clock skew` mean in this context. Also, `iat`, `nbf` and `exp`, while in the spec, need some expansion here.

Is this maybe for a separate documentation page? (`README.md` should be short and to the point)

lcobucci replied:
> Is this maybe for a separate documentation page? (README.md should be short and to the point)

I agree with you @Ocramius but let's do that as a separate PR, we can keep it on the `README.md` for now and modify it later.

Ocramius commented on README.md:
```diff
@@ -88,6 +88,27 @@ var_dump($token->validate($data)); // true, because current time is between "nbf
+
+var_dump($token->validate($dataWithLeeway)); // false, because token can't be used before now() + 60 and it's now() + 14
+
+$dataWithLeeway->setCurrentTime(time() + 51); // changing the validation time to future
```
Use `$time = time()`, assigned only once to make it clear that we're working within the boundaries of fixed time throughout the example

Ocramius commented on README.md:
````diff
@@ -97,6 +118,11 @@ var_dump($token->validate($data)); // false, because token is expired since curr
 configured in ```ValidationData``` they will be ignored by ```Token::validate()```.
 - ```exp```, ```nbf``` and ```iat``` claims are configured by default in ```ValidationData::__construct()```
 with the current UNIX time (```time()```).
+- The values for the keys in the  optional ```$leeway``` array of ```ValidationData``` will add that number of seconds 
````
Same here about `$leeway`. A link is required

Ocramius commented on src/ValidationData.php:
```diff
@@ -24,24 +24,28 @@ class ValidationData
      */
     private $items;
 
+    private $timeFields = ['iat', 'nbf', 'exp'];
```
Docblock with `@var string[]` needed

Ocramius commented on src/ValidationData.php:
```diff
@@ -24,24 +24,28 @@ class ValidationData
 
+    private $timeFields = ['iat', 'nbf', 'exp'];
+
+    private $leeways = [];
```
Docblock with types needed

Ocramius commented on src/ValidationData.php:
```diff
@@ -24,24 +24,28 @@ class ValidationData
      * Initializes the object
      *
      * @param int $currentTime
+     * @param array $leeways
```
Please use `type[]` (with the correct type) here

Ocramius replied:
This kind of definition of leeways is a bit squishy and hard to keep in mind, as the user needs to know the keys of the `$leeways` parameters. An explicit parameter for each of the fields supporting `$leeways` is more explicit and avoids basic mistakes.

Ocramius commented on src/ValidationData.php:
```diff
@@ -91,9 +95,13 @@ public function setSubject($subject)
-        $this->items['nbf'] = (int) $currentTime;
-        $this->items['exp'] = (int) $currentTime;
+        foreach ($this->timeFields as $field) {
+            if (isset($this->leeways[$field]) && is_int($this->leeways[$field])) {
```
The `is_int()` is to be avoided here - use an assertion where `$this->leeways` is being modified instead, and throw if a non-int is passed in.

Ocramius commented on test/unit/ValidationDataTest.php:
```diff
@@ -209,16 +279,18 @@ private function createExpectedData(
+            'iss' => $iss !== null ? (string)$iss : null,
+            'aud' => $aud !== null ? (string)$aud : null,
+            'sub' => $sub !== null ? (string)$sub : null,
+            'iat' => (int)$iat,
```
This cast is likely wrong - should never come in as non-integer here

Ocramius commented on test/unit/ValidationDataTest.php:
```diff
@@ -200,7 +268,9 @@ public function claimValues()
      * @param string $aud
-     * @param int $time
+     * @param int $iat
+     * @param int $nbf
```
`int|null`, I suppose?

Ocramius commented on test/unit/ValidationDataTest.php:
```diff
@@ -200,7 +268,9 @@ public function claimValues()
-     * @param int $time
+     * @param int $iat
+     * @param int $nbf
+     * @param int $exp
```
`int|null`, I suppose?

m777z commented:
Thanks for the feedback, guys! Since some pretty big changes are needed, should I continue to send commits to this pull request or just close it and open a new PR when the code is ready?

Ocramius commented:
Sending patches here would give us an idea if thr comments are addressed:
incremental reviews FTW!

On Sat, 7 Jul 2018, 07:02 m777z, <notifications@github.com> wrote:

> Thanks for the feedback, guys! Since some pretty big changes are needed,
> should I continue to send commits to this pull request or just close it and
> open a new PR when the code is ready?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/lcobucci/jwt/pull/248#issuecomment-403189069>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAJakORKzJ6Gn3ws90uQCzWYho1fq4Boks5uEEDvgaJpZM4U-Iw1>
> .
>

m777z commented:
@lcobucci I updated the code to only take one integer parameter for the leeway (which is then applied to each of the time-based checks).

@Ocramius The code style changes were reverted. I made some changes to README.md (used your $time = time() suggestion), though as you and @lcobucci said we may eventually want to link to a good explanation elsewhere since I am having trouble succinctly explaining leeway and clock skew. I believe all your comments on src/ValidationData.php have been indirectly addressed by the simplification I made on @lcobucci's suggestion. I also added the |null in test/unit/ValidationDataTest.php and took out the unneeded cast.

Please let me know what further changes need to be made.

lcobucci commented:
@m777z awesome, thanks for your work!

This week I'll be releasing the last patch on 3.2.x and focus on 3.3, thus your code will be reviewed ASAP

lcobucci commented on src/ValidationData.php:
```diff
@@ -24,24 +24,30 @@ class ValidationData
+     * The leeway (in seconds) to use when validating time claims
+     * @var int
+     */
+    private $leeway;
```
I was wondering if it makes sense to have this as a property... I mean, it's only used for configuring the current time, maybe it would be better to simply pass it to `ValidationData#setCurrentTime()`... @Ocramius any thoughts?

m777z replied:
I'm completely fine with doing it either way. Leaving it as it is makes sense if it's fairly common to use the `setCurrentTime()` method 1 or more times after initial construction of the object while keeping the same leeway. However, if users would want to keep changing the leeway without constructing a new object, or if they rarely use `setCurrentTime()` directly at all, then passing leeway as a parameter to `setCurrentTime()` makes more sense.

lcobucci replied:
@m777z I'm have no clue on how users are using this object, and since it will be deprecated very soon I don't think we should debate this much. Let's go with your suggested approach and if needed we change it. Thanks!

m777z replied:
@lcobucci I am assuming you mean it is okay to just leave things as they are for this part. If so, are there other changes I need to make before this pull request is merged?

lcobucci replied:
@m777z that's correct. There're some small CS changes (which I'll handle ASAP), I just didn't manage to work on this. Plan to work on it tomorrow

lcobucci commented on test/unit/ValidationDataTest.php:
```diff
@@ -209,16 +238,18 @@ private function createExpectedData(
-            'nbf' => $time,
-            'exp' => $time
+            'iat' => $iat,
+            'nbf' => $nbf !== null ? $nbf: $iat,
```
It's sad that we can't use `??` to simplify this due to PHP 5.x compatibility :sob:

lcobucci commented:
@m777z thanks for your contribution (and sorry about the delay!) :smile:
